### DISCUSSION TITLE
Keycloak Realm Configuration Management Tools

### DISCUSSION THREAD

**Discussion by thomasdarimont** [keycloak/keycloak#30643] [Feedback]
Numerous options exist for managing Keycloak Realm configurations within the 
Keycloak ecosystem. As the Keycloak team, we want to understand better what works 
best for the community and how we can improve the support for Realm configuration 
management.

The options that exist on our radar are as follows:
- [CNF1 Keycloak Admin CLI kcadm.sh](https://www.keycloak.org/docs/latest/server_admin/#admin-cli)
- [CNF2 Keycloak-Config-CLI](https://github.com/adorsys/keycloak-config-cli)
- [CNF3 Terraform Provider for Keycloak](https://github.com/mrparkers/terraform-provider-keycloak)
- [CNF4 Keycloak Ansible](https://github.com/ansible-middleware/keycloak)
- [CNF5 Keycloak Pulumi](https://www.pulumi.com/registry/packages/keycloak/)
- [CNF6 Crossplane Provider for Keycloak](https://github.com/crossplane-contrib/provider-keycloak)
- [CNF7 Keycloak JSON Import / Export](https://www.keycloak.org/server/importExport)
- [CNF8 Keycloak Operator Realm Import via Custom Resources](https://www.keycloak.org/operator/advanced-configuration)
- CNF9 Hand-made Realm Configuration Management
- CNF10 Another approach: Please specify

So that we in the Keycloak community have a representative picture of the configuration 
options used, we would also like you to participate in the following [following brief, anonymous survey](https://forms.gle/2PtNt9hruCwwDm5P8).

Feel free to add additional comments or suggestions regarding realm configuration management tooling below!

Thank you for your help!

**Comment by rene6502** [comment]
Thanks for bringing this up.\
I consider this a critical topic since automatic Keycloak configuration is essential for many projects.\
Developing proprietary tools for this purpose is time-consuming, and relying on third-party tools can be risky if they are not well-maintained.\
Consider a scenario where you need to update Keycloak to incorporate new features or security enhancements,\
but the tool you've been using does not support the latest version, and there's uncertainty about if or when it will be updated.


Here are some requirements I envision for such a configuration tool:
- The configuration tool should perform a one-way synchronization from a human-readable, file-based configuration to Keycloak.
- It must ensure that after each run, the configuration in Keycloak precisely matches the input configuration.
- The tool should not assume that resources remain unchanged by other actors.
- It must not delete resources that it did not create.\
  However, it should be capable of updating resources created by others.
- It must not delete a realm unless explicitly instructed to do so.
- It should support all resource types, including users.
- The tool should support YAML as the configuration file format for its readability over JSON and compatibility with Helm as a robust templating engine.
- It should allow the configuration to be divided into multiple files, for instance, separate files for each client.
- It must support default values; the configuration files can omit values that have defaults, which must be documented.
- The tool should be compatible with Kubernetes, Docker, and local standalone installations.
- It should permit the inclusion of secrets, such as through environment variables.
- The tool must log all changes, including resources added, deleted, or updated.

Export
- The export order of elements should be deterministic.
- Optionally remove internal data, such as generated IDs or timestamps.
- I want to export a realm and compare it with a previous export.\
  The diff should only display genuine changes.\
  This should work even when the Keycloak database was created from scratch.

**Reply by tanadeau** [reply]
I believe keycloak-config-cli has most of these features. I wish it could be brought into Keycloak's org so that there can be simultaneous releases, continuous testing, etc. If not, if there could be a way to coordinate releases so the CLI's release isn't offset by several weeks for new major versions.

**Reply by jaehnigen-kp** [reply]
I'm pretty sure that there's no one-size-fits-all solution. The project's ecosystems for example in terms of tooling and knowledge are just different. We also don't see a need for a ("the") new tool, but for proper maintenance of the existing ones. And some of them are already pretty awesome!

**Reply by amiv1** [reply]
> The tool should support YAML as the configuration file format for its readability over JSON and compatibility with Helm as a robust templating engine.

I wouldn't make this a hard requirement though. For example, Terraform has its own DSL and it is one of industry standards for managing configuration as code. I wouldn't remove it from considerations just because the declaration format isn't YAML.

**Reply by algra4** [reply]
Hello,

Another way to configure Keycloak is to use [Keycloak Admin CLI](https://mvnrepository.com/artifact/org.keycloak/keycloak-admin-cli) as explained by @MaikKingma during the DevoxxBe https://youtu.be/aZOoH0i4s-0?si=6dWLyZo0ynrLEofR

**Reply by danielshiplett** [reply]
I agree with all of these points except the YAML/Helm.  Like @gim- mentioned, Terraform HCL is a human readable and declarative language too.

Also, not everything runs on Kubernetes.  So, Helm isn't a hard requirement.  Many people still run software in other forms and don't/wouldn't use Helm to deploy their configuration.

**Reply by ace130-github** [reply]
I would like to add two requirements:
* The tool must be configurable by environment variables at its runtime (this is certainly implied by @rene6502 's requirements, but I think it deserves an explicit mention).
* Running the tool with a given configuration should be idempotent.

**Comment by jonasvoelcker** [comment]
Hey there,

on our project we suffered quite long from documenting changes in Confluence which our Ops-Team had to apply to all our stages and also we were not able to keep the dev stages clean from manual changes or missed configurations.

When we started the config as code approach we took our current realm export and wanted to make it work. We didn't use terraform or other providers as this would have required a lot of further implementation.

Now it is possible to add a new flow, client or configuration at our dev stages which turns into the same output we use as input for the keycloak-config-cli and with some tiny adjustments we can adapt the same parts in our realm configuration, which is configured using environment variables.

It definitely is a bunch of work when a new major comes in. But honestly the work is easy to manage, the more people participate in the task. The last majors we had several people giving ideas, approaches, the right links to bug descriptions and so on and therefore it was manageable to adjust the tool.

As every project ends somewhere, a larger base of contributors always gives certainty that a tool survives 😉

**Reply by tanadeau** [reply]
And your tool is very much appreciated 😁

**Reply by jaehnigen-kp** [reply]
As we already had the general Terraform knowledge, that's exactly why we opted for the Terraform provider approach: all the benefits of a declarative, coded approach: easy to review, version-controllable, repeatable (e.g. for staging) and so on.

**Reply by timadevelop** [reply]
We switched from tf to keycloak-config-cli and it's been an amazing experience. I genuinely hope this tool gets more attention from the community

**Comment by Breee** [comment]
- I'm maintainer of [CNF6 Crossplane Provider for Keycloak](https://github.com/crossplane-contrib/provider-keycloak), it currently directly depends on  [CNF3 Terraform Provider for Keycloak](https://github.com/mrparkers/terraform-provider-keycloak) due to the nature of how  Upjet generated providers and Crossplane work (Also because i cannot provide a full selfwritten provider as a single person)

- For me  [CNF3 Terraform Provider for Keycloak](https://github.com/mrparkers/terraform-provider-keycloak) has the top 1 priority, not only because my project depends on it, but also because there are alot of companies out that that rely on it and also on its integrity. 
- IMO we as community should and must find a way to keep maintaining it / improving it, as Identity and Access Management as well as SSO is a critical part of security. In a cloud native world it just makes sense to have IaC tools that can store their configuration in Git as single source of truth to eliminate config drifts and unapproved granting of permissions. This has been a best practice for **years**.

**Comment by schuerg** [comment]
Improvements to the [CNF7 Keycloak JSON Import / Export](https://www.keycloak.org/server/importExport) would be beneficial in my humble opinion:

1. Allow exporting full Keycloak realm JSON files during runtime without the need to shut down Keycloak. This should include full realm.json exports, with users, hashed credentials, and keys, not just partial exports. This would also be a great way doing backups without downtime.
2. Allow exporting Keycloak realms via the Admin REST API for users with a specific role. Something like the now unmaintained community extension *keycloak-export*[^2]
3. Enable the import of Keycloak realm.json files using the JSON-Patch format[^3] or something similar. This would facilitate fine-grained operations on the entire realm JSON structure, such as Add, Remove, Replace, Copy, Move, and Test of JSON object attributes and JSON array elements, similar to the functionality provided by the Kubernetes tool Kustomize.


---

[^2]: https://github.com/cloudtrust/keycloak-export
[^3]: https://jsonpatch.com/

**Comment by NeckBeardPrince** [comment]
Correct me if I'm wrong here, but don't the [Pulumi](https://github.com/search?q=repo%3Apulumi%2Fpulumi-keycloak%20mrparker&type=code) and [Crossplane](https://github.com/search?q=repo%3Acrossplane-contrib%2Fprovider-keycloak%20mrparker&type=code) providers also rely on the terraform module created by @mrparkers?

**Reply by thomasdarimont** [reply]
Yes, you are right. However, for this survey we want to learn what tools users use to configure Keycloak. 
If you use pulumi, then that's your interface and the keycloak terraform provider is "just" an implementation detail.

**Reply by NeckBeardPrince** [reply]
Makes sense, thank you.

**Comment by pauliuspetk** [comment]
I've just set up a simple Keycloak environment for the first time, using the two-stage Docker approach, as per the documentation. I only use it as an IDP broker between my application and external auth providers.

I wish `kc.sh export` would be able to fully export realm configurations at runtime. (Disallowing import/export of the master realm is fine for me (I think?), and excluding users is also fine, I just need the other realm(s) configurations, incl. client settings.) This would allow me to simply do a `docker exec` (or ECS exec, etc.) inside a container with a running KC instance and get the .json(s), without needing to start a new container or stop existing ones.

I'd also like sensitive stuff like client secrets to be symmetrically encrypted using some encryption key stored in the keystore.

Similarly, I'd like `kc.sh start --import-realms` to support the `--override` flag so that I could just import the realms exported as above, and that sensitive values would get decrypted using the key in the keystore.

This together would allow me to safely version realm configs in a private git repo without any third-party tooling asides from `docker` and `git` and would make deploying new KC environments a breeze.

It seems like `--import-realms` deletes whatever is in the import directory right after ingesting it and writing it to the DB, which would work out fine to avoid unnecessary re-imports. I'd only need to mount a shared directory (or EFS, etc.) on `/opt/keycloak/data/import/` and drop the .json(s) there as I see fit.

The only currently available native import/export options for Docker-based deployments seem to be CFN1 and CFN7, all other require some external dependencies or tooling in one form or another, that's maintained separately, and has no guarantee to stay up to date.

The CFN3 Terraform provider is looking like least complex option that would work for my simple use case that allows me to version realm configurations while storing the sensitive values somewhere secure. But it adds quite a bit of complexity that could feasibly be avoided, and I'd much rather have a simpler native option available similar to what is described above using the built-in `kc.sh` or `kcadm.sh`.

**Comment by MarcScheib** [comment]
I have created https://github.com/CycriLabs/keycloak-configurator (therfore fitting into CNF9) which I use for my personal projects and which we use in our company. It is based on Quarkus together with picocli and can be executed from the CLI or via docker. Internally, it uses keycloak admin client to configure keycloak based on separate JSON files for each entity type (e.g. realms, clients, users, groups, roles, etc.).

Besides keycloak configuration, it supports exporting it again, exporting secrets / env files (using velocity templates), and secret rotation.

Currently, we use it for the initial configuration of Keycloak in productions servers as well as for configuring short-living test servers or feature servers based on a common shared configuration repository. Whenever there is config change, we simply add the related JSON file and run the tool (i.e. partial updates/rolling updates, currently no sync is supported). 

In addition, we use it for integration testing as well. We do not need to maintain a large realm.json for startup import but instead use fine-grained single files for better overview and source management.

Looking forward for feedback, feature requests, or anything else that can improve the tool.

**Reply by tbroyer** [reply]
Interesting. How does it compare to keycloak-config-cli? (besides the config files layout, IIUC)

**Reply by MarcScheib** [reply]
I am not very familiar with keycloak-config-cli, but on first sight it looks very similar in regards to configuration. Keycloak-config-cli has for sure more config features and is more mature.

Not sure though about other features like export, secret export/rotation, etc. 
Underlying frameworks are different as well (spring vs quarkus). 

Keycloak-configurator evolved out of our previous python-based solution. It was some years old and maintance was an issue. Therefore, we decided to migrate the features to a Java tech stack.

**Reply by bencat-sixense** [reply]
keycloak-config-cli can Import, but doesn't Export, so I'll try yours asap :)
Thank you @MarcScheib !

**Comment by ravthiru** [comment]
For our project we use combination of both 

CNF1 Keycloak Admin CLI kcadm.sh 
CNF2 Keycloak-Config-CLI 

Was also wondering if there is way to sync any new changes to keycloak instance to git using a pull request. 
So that we dont missout on the fly configurations done to the keycloak instance.

**Comment by b-thiswatch** [comment]
Hi,

in our project we are using IaC on K8s via ArgoCD. So for us improvement to [CNF8 Keycloak Operator Realm Import via Custom Resources](https://www.keycloak.org/operator/advanced-configuration) would be the best option. 

The improvement being that it is not just "Realm Import via CR", but full management via CR. 
Having an optionally immutable realm (via UI) that can just be managed via CR, would also be a nice feature.

**Comment by danielshiplett** [comment]
Is there a timeframe in which this discussion and the survey are going to close out?

My team is making use of both [CNF2 Keycloak-Config-CLI](https://github.com/adorsys/keycloak-config-cli) and [CNF3 Terraform Provider for Keycloak](https://github.com/mrparkers/terraform-provider-keycloak).

I would prefer to use [CNF3 Terraform Provider for Keycloak](https://github.com/mrparkers/terraform-provider-keycloak) since Terraform is a common pattern for a lot of other infrastructure my team manages.  It also gives us a free audit trail by running Terraform through GitOps.

**Comment by fpoyer-eurecia** [comment]
I know this thread is somewhat old, but I am to choose an solution for my company, so I'd be more than interested in a follow up, or at least the results of the initial survey.

Given the severe limitations of CNF7 (JSON import) and CNF8 (K8s Operator) that can only import if the realm doesn't exist, or delete and recreate it (so no update, which is real bad for production environment), CNF2 (Keycloak-Config-CLIT) and CNF3 (Terraform Provider) seem like the best options, as both use a declarative approach and can scale even for complex projects and can manage updates of existing ressources.

But neither is an "official" tool, implying at least some delay between a KC release and the corresponding update (for example : neither seem to be handling the recently added Organization concept, at least from documentation), and though both projects seem pretty active at the moment, one may worry about maintenance in the future.

At least having the result of the survey could give some visibility into what the community is using, and bit of confidence about what tool should survive...

**Reply by NeckBeardPrince** [reply]
Are aware that the terraform provider was adopted by Keycloak? https://github.com/keycloak/terraform-provider-keycloak

**Reply by fpoyer-eurecia** [reply]
I was not, thanks for the heads up !
While searching with Google and asking IAs, I was always redirected to mrparkers/keycloak (https://registry.terraform.io/providers/mrparkers/keycloak/latest/docs).

That's settled then Terraform it is !

**Comment by DeabitTech** [comment]
yeah, i'm building a simple bitnami keycloak instance using docker, so i setup docker file and docker compose and launch the compose.

but i'm in trouble about to setup the realm that i'm importing as  "current realm"  on the bootstrap of the instance, so in the build time.

Is there a way to do that?

Thanks in advance
